hive/lib/src/binary/binary_reader_impl.dart

```diff
@@ -243,13 +243,14 @@ class BinaryReaderImpl extends BinaryReader {
   /// Not part of public API
   Frame? readFrame(
       {HiveCipher? cipher, bool lazy = false, int frameOffset = 0}) {
+    // frame length is stored on 4 bytes
     if (availableBytes < 4) return null;
 
+    // frame length should be at least 8 bytes
     var frameLength = readUint32();
-    if (frameLength < 8) {
-      throw HiveError(
-          'This should not happen. Please open an issue on GitHub.');
-    }
```

Comment:
The check remains the same however now it return `null` like other checks.

This check isn't different from other checks.
```dart
// 1) frame length is stored on 4 bytes
if (availableBytes < 4) return null;

// 2) frame length should be at least 8 bytes
var frameLength = readUint32();
if (frameLength < 8) return null;

// 3) frame is bigger than avaible bytes
if (availableBytes < frameLength - 4) return null;

// 4) frame is corrupted or provided chiper is different
if (computedCrc != crc) return null;
```

Let me explain:
## 1) frame length is stored on 4 bytes
Imagine the following situation we want to put something in the box: `await box.put(frame1, frame1);`
So Hive would write `[27, 0, 0, 0, 1, 6, 102, 114, 97, 109, 101, 49, 4, 6, 0, 0, 0, 102, 114, 97, 109, 101, 49, 13, 235, 12, 102]`
And the OS could only write for example first three bytes `[27, 0, 0]` and the rest were cut off

**This frame is invalid** because we know the first four bytes represents the length ( _note: length counts itself too so it is at least 4 bytes long_ )

## 2) frame length should be at least 8 bytes
Imagine the following situation we want to put something in the box: `await box.put(frame1, frame1);`
So Hive would write `[27, 0, 0, 0, 1, 6, 102, 114, 97, 109, 101, 49, 4, 6, 0, 0, 0, 102, 114, 97, 109, 101, 49, 13, 235, 12, 102]`
And the OS mess up the first byte `[7, ...]` or we calculate it wrong or old Hive version would write `[0, 0, 0, 0]`

**This frame is invalid** because a frame is at least 8 bytes ( `bytes means length` ) since the Frame length is already consumes 4 bytes and CRC also consumes 4 bytes not to mention the key / value and their types also consumes bytes.

## 3) frame is bigger than avaible bytes
Imagine the following situation we want to put something in the box: `await box.put(frame1, frame1);`
So Hive would write `[27, 0, 0, 0, 1, 6, 102, 114, 97, 109, 101, 49, 4, 6, 0, 0, 0, 102, 114, 97, 109, 101, 49, 13, 235, 12, 102]`
And the OS mess up the first byte `[28, ...]` or we calculate it wrong or OS cut off some bytes

**This frame is invalid** because we want to read 28 bytes instead of 27 bytes and there is no enough space OR in case where there are multiple frames side by side then we read from next frame and CRC check will yield error ( `null` ).

## 4) frame is corrupted or provided chiper is different
Imagine the following situation we want to put something in the box: `await box.put(frame1, frame1);`
So Hive would write `[27, 0, 0, 0, 1, 6, 102, 114, 97, 109, 101, 49, 4, 6, 0, 0, 0, 102, 114, 97, 109, 101, 49, 13, 235, 12, 102]`
And the OS mess up for instance the middle byte or CRC itself

**This frame is also invalid**

There is no difference between these errors.

And if you search for `BinaryReaderImpl.readFrame` references everywhere `null` is handled properly.
[When normal Box loads](https://github.com/hivedb/hive/blob/v2.0.5/hive/lib/src/binary/frame_helper.dart#L26) it is used to determine `recoveryOffset`

[When lazyBoy loads it's keys](https://github.com/hivedb/hive/blob/v2.0.5/hive/lib/src/io/frame_io_helper.dart#L78) it is used to determine `recoveryOffset`

[When lazyBoy reads value](https://github.com/hivedb/hive/blob/v2.0.5/hive/lib/src/backend/vm/storage_backend_vm.dart#L112) it is handled properly however we would never reach here because when `lazyBox` loads it's keys ( _above_ ) it runs the checks.

Reply:
And again the first commit contains the main fix: `await writeRaf.setPosition(recoveryOffset);`

I recommend to try only **2) frame length should be at least 8 bytes** fix.
You can play around with my example code first of all comment out the first commit's fix ( `// await writeRaf.setPosition(recoveryOffset);` ) and now example code results double box recovery when you run my example. Change `openBox` to `openLazyBox` or corrupt the first frame it should always work.

Reply:
Yeah I got the corruption part. Anyways nevermind.